testcontainers/testcontainers-dotnet
# feat: Support digest in `IImage`, `DockerImage` and in the `WithImage(string)` implementation

## What does this PR do?



Support for digest passed to the image parser.
Dropped workaround to fully utilize docker api and support digest in the requirements.

## Why is it important?



OpenTelemetry recommend to reference all docker images by the name, tag, followed by the digest.
Part of implementing this recommendation is https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3514

Passing full name to the Container API leads to

```log
System.ArgumentException
Cannot parse image: confluentinc/cp-zookeeper:7.6.1@sha256:4dc780642bfc5ec3a2d4901e2ff1f9ddef7f7c5c0b793e1e2911cbfb4e3a3214 (Parameter 'image')
   at DotNet.Testcontainers.Images.MatchImage.Match(String image) in /_/src/Testcontainers/Images/MatchImage.cs:line 34
   at DotNet.Testcontainers.Images.DockerImage..ctor(String image) in /_/src/Testcontainers/Images/DockerImage.cs:line 41
   at DotNet.Testcontainers.Builders.ContainerBuilder`3.WithImage(String image) in /_/src/Testcontainers/Builders/ContainerBuilder`3.cs:line 67
   at IntegrationTests.KafkaFixture.LaunchZookeeper(INetwork containerNetwork) in C:\GitHub\opentelemetry-dotnet-instrumentation\test\IntegrationTests\KafkaCollection.cs:line 63
   at IntegrationTests.KafkaFixture.InitializeAsync() in C:\GitHub\opentelemetry-dotnet-instrumentation\test\IntegrationTests\KafkaCollection.cs:line 39
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90
```


## Related issues



- Closes #1214
- Relates https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3514
- Potentially impacts https://github.com/dotnet/Docker.DotNet/issues/595

## How to test this PR


Unit tests extended.
You can check also code similar to.

```csharp
        var container = new ContainerBuilder()
            .WithImage("confluentinc/cp-zookeeper:7.6.1@sha256:4dc780642bfc5ec3a2d4901e2ff1f9ddef7f7c5c0b793e1e2911cbfb4e3a3214")
            .Build();
```

Do not tested with Docker 4..13.1. The latest version is fine. The question is if the support can be dropped for 2-years old version. I think that it can be dropped as Docker for Desktop was released couple of time with security fixes. 4.13.1 should be treated as insecure and not recommended to use.

## Follow-ups



Release new version. I will be happy to proceed with https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3514
